perllib/FixMyStreet/Cobrand/Smidsy.pm

```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
     }
 }
 
+sub loop_period {
+    my ($date, $period, $end, $extra) = @_;
+    $end = FixMyStreet->set_time_zone($end)->truncate(to => $period)->add($period . 's' => 1)->subtract(seconds => 1);
+    my @out;
+    while ($date <= $end) {
+        push @out, { n => $date->$period, $extra ? (d => $date->$extra) : () };
+        $date->add($period . 's' => 1);
+    }
+    return @out;
+}
+
+sub get_body_children {
+    my ($self, $body) = @_;
+
+    $body = $self->{c}->model('DB::Body')->find( { id => $body } );
+    my $children = $body->first_area_children;
+    $self->{c}->stash->{area_body_name} = $body->name;
+    return [ values %$children ];
+}
+
+sub report_page_data {
+    my $self = shift;
+    my $c = $self->{c};
+
+    if (my $body = $c->get_param('body')) {
+        $body = $c->model('DB::Body')->find( { id => $body } );
+        if ($body) {
+            $body = $c->cobrand->short_name($body);
+            $c->res->redirect("/reports/$body");
+            $c->detach;
+        }
+    }
+
+    my $start = DateTime->new(year => 2013, month => 1, day => 1);
+    my $end = DateTime->now();
+
+    $c->stash->{start_date} = $c->get_param('start_date') || '2013-01-01';
+    $c->stash->{end_date} = $c->get_param('end_date');
+    $c->forward('/dashboard/construct_rs_filter') if $c->stash->{start_date};
+
+    if ( my $source = $c->get_param('sources' ) ) {
+        my $rs = $c->stash->{problems_rs};
+        if ( $source ne 'stats19' ) {
+            $source = undef;
+        }
+        $c->stash->{problems_rs} = $rs->search( { external_body => $source } );
+    }
+
+    my %participants = (
+        bicycle => 'Bicycle',
+        bus => 'Bus',
+        car => 'Car',
+        hgv => 'HGV',
+        horse => 'Horse',
+        minibus => 'Minibus',
+        motorcycle => 'Motorcycle',
+        pedestrian => 'Pedestrian',
+        taxi => 'Taxi',
+        tram => 'Tram',
+        van => 'Van',
+    );
+
+    $c->stash->{participant_labels} = \%participants;
+
+    if ( my $transport = $c->get_param('transport') ) {
+        if ( $participants{$transport} ) {
+            $c->stash->{transport} = $transport;
+            my $rs = $c->stash->{problems_rs};
+            my $filter = "SIMILAR TO '%T12:participants,T\\d+:$transport,%'";
+            $c->stash->{problems_rs} = $rs->search( { extra => \$filter } );
```

Comment:
This is a SQL injection. You need to use placeholders as otherwise it is not escaped. See https://metacpan.org/pod/SQL::Abstract#Literal-SQL-with-placeholders-and-bind-values-(subqueries)